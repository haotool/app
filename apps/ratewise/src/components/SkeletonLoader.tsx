/**
 * Skeleton Loader Component
 * [Lighthouse-optimization:2025-10-28] 取代 loading spinner，提升 perceived performance
 *
 * 參考:
 * - https://blog.logrocket.com/ux-design/skeleton-loading-screen-design/
 * - https://react.dev/reference/react/Suspense
 *
 * 優勢:
 * - 即時顯示內容結構，減少 perceived loading time
 * - 用戶可預期內容佈局，降低 cognitive load
 * - 避免 spinner 的「等待感」，提升 UX
 */

export const SkeletonLoader = () => {
  return (
    <div className="min-h-screen bg-slate-50 p-4 md:p-6" role="status" aria-live="polite">
      <div className="mx-auto max-w-4xl space-y-6">
        {/* Header Skeleton */}
        <div className="animate-pulse">
          <div className="h-8 w-48 bg-slate-200 rounded-lg mb-2"></div>
          <div className="h-4 w-32 bg-slate-200 rounded"></div>
        </div>

        {/* Mode Switcher Skeleton */}
        <div className="flex gap-2 mb-4 animate-pulse">
          <div className="h-10 w-32 bg-slate-200 rounded-lg"></div>
          <div className="h-10 w-32 bg-slate-200 rounded-lg"></div>
        </div>

        {/* Main Converter Card Skeleton */}
        <div className="bg-white rounded-2xl shadow-lg p-6 space-y-4 animate-pulse">
          {/* Currency Selectors */}
          <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
            <div className="space-y-2">
              <div className="h-4 w-16 bg-slate-200 rounded"></div>
              <div className="h-12 bg-slate-200 rounded-lg"></div>
            </div>
            <div className="space-y-2">
              <div className="h-4 w-16 bg-slate-200 rounded"></div>
              <div className="h-12 bg-slate-200 rounded-lg"></div>
            </div>
          </div>

          {/* Amount Input */}
          <div className="space-y-2">
            <div className="h-4 w-20 bg-slate-200 rounded"></div>
            <div className="h-14 bg-slate-200 rounded-lg"></div>
          </div>

          {/* Result Display */}
          <div className="bg-purple-50 rounded-xl p-4 space-y-2">
            <div className="h-6 w-full bg-purple-200 rounded"></div>
            <div className="h-4 w-3/4 bg-purple-200 rounded"></div>
          </div>

          {/* Quick Amounts */}
          <div className="flex gap-2 flex-wrap">
            {[1, 2, 3, 4].map((i) => (
              <div key={i} className="h-10 w-20 bg-slate-200 rounded-lg"></div>
            ))}
          </div>
        </div>

        {/* Trend Chart Skeleton */}
        <div className="bg-white rounded-2xl shadow-lg p-6 animate-pulse">
          <div className="h-4 w-32 bg-slate-200 rounded mb-4"></div>
          <div className="h-64 bg-slate-200 rounded-lg"></div>
        </div>

        {/* Favorites Skeleton */}
        <div className="bg-white rounded-2xl shadow-lg p-6 animate-pulse">
          <div className="h-4 w-24 bg-slate-200 rounded mb-4"></div>
          <div className="grid grid-cols-2 md:grid-cols-3 gap-2">
            {[1, 2, 3].map((i) => (
              <div key={i} className="h-12 bg-slate-200 rounded-lg"></div>
            ))}
          </div>
        </div>
      </div>
      <span className="sr-only">載入匯率資料中...</span>
    </div>
  );
};

/**
 * Lightweight Currency Card Skeleton
 * 用於貨幣列表的輕量化骨架屏
 */
export const CurrencyCardSkeleton = () => {
  return (
    <div className="bg-white rounded-lg p-3 shadow-sm animate-pulse" role="status">
      <div className="flex items-center justify-between">
        <div className="flex items-center gap-2">
          <div className="w-8 h-8 bg-slate-200 rounded-full"></div>
          <div className="space-y-1">
            <div className="h-4 w-12 bg-slate-200 rounded"></div>
            <div className="h-3 w-20 bg-slate-200 rounded"></div>
          </div>
        </div>
        <div className="h-4 w-16 bg-slate-200 rounded"></div>
      </div>
    </div>
  );
};

/**
 * Converter Mode Skeleton
 * 針對特定轉換模式的骨架屏
 */
export const ConverterSkeleton = ({ mode }: { mode: 'single' | 'multi' }) => {
  if (mode === 'multi') {
    return (
      <div className="space-y-4 animate-pulse" role="status">
        <div className="h-12 bg-slate-200 rounded-lg"></div>
        <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
          {[1, 2, 3, 4].map((i) => (
            <CurrencyCardSkeleton key={i} />
          ))}
        </div>
      </div>
    );
  }

  return (
    <div className="space-y-4 animate-pulse" role="status">
      <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
        <div className="h-14 bg-slate-200 rounded-lg"></div>
        <div className="h-14 bg-slate-200 rounded-lg"></div>
      </div>
      <div className="h-16 bg-purple-200 rounded-xl"></div>
    </div>
  );
};
