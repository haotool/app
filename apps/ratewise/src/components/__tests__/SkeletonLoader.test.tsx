/**
 * SkeletonLoader Test Suite
 *
 * 測試策略 (Linus Style):
 * 1. 測試行為，不測實作細節 (test behavior, not implementation)
 * 2. 用戶視角測試 (user-centric testing with ARIA)
 * 3. 簡潔優先 (simple, focused tests)
 */

import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';
import { SkeletonLoader, CurrencyCardSkeleton, ConverterSkeleton } from '../SkeletonLoader';

describe('SkeletonLoader', () => {
  describe('基本渲染', () => {
    it('渲染完整頁面骨架屏', () => {
      render(<SkeletonLoader />);

      // ✅ Accessibility: role="status" 正確設置
      const loader = screen.getByRole('status');
      expect(loader).toBeInTheDocument();

      // ✅ SR-only 文字存在
      expect(screen.getByText('載入匯率資料中...')).toBeInTheDocument();
    });

    it('渲染所有主要區塊', () => {
      const { container } = render(<SkeletonLoader />);

      // ✅ 使用快照測試驗證結構穩定性
      // Linus 原則: "如果結構改變，測試應該失敗"
      expect(container.firstChild).toMatchSnapshot();
    });

    it('應用正確的動畫類別', () => {
      const { container } = render(<SkeletonLoader />);

      // ✅ 檢查 animate-pulse 存在於多個區塊
      const animatedElements = container.querySelectorAll('.animate-pulse');
      expect(animatedElements.length).toBeGreaterThan(0);
    });
  });

  describe('Accessibility', () => {
    it('包含 aria-live="polite" 屬性', () => {
      const { container } = render(<SkeletonLoader />);

      const liveRegion = container.querySelector('[aria-live="polite"]');
      expect(liveRegion).toBeInTheDocument();
    });

    it('包含螢幕閱讀器專用文字', () => {
      render(<SkeletonLoader />);

      const srText = screen.getByText('載入匯率資料中...');
      expect(srText).toHaveClass('sr-only');
    });
  });
});

describe('CurrencyCardSkeleton', () => {
  it('渲染貨幣卡片骨架', () => {
    render(<CurrencyCardSkeleton />);

    // ✅ role="status" 存在
    expect(screen.getByRole('status')).toBeInTheDocument();
  });

  it('包含正確的結構元素', () => {
    const { container } = render(<CurrencyCardSkeleton />);

    // ✅ 圓形頭像 skeleton
    const avatar = container.querySelector('.rounded-full');
    expect(avatar).toBeInTheDocument();

    // ✅ 動畫效果
    expect(container.querySelector('.animate-pulse')).toBeInTheDocument();
  });

  it('匹配設計快照', () => {
    const { container } = render(<CurrencyCardSkeleton />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('ConverterSkeleton', () => {
  describe('模式: single', () => {
    it('渲染單一轉換器骨架', () => {
      render(<ConverterSkeleton mode="single" />);

      const skeleton = screen.getByRole('status');
      expect(skeleton).toBeInTheDocument();
    });

    it('不渲染 CurrencyCardSkeleton', () => {
      const { container } = render(<ConverterSkeleton mode="single" />);

      // ✅ single 模式沒有圓形頭像
      const avatars = container.querySelectorAll('.rounded-full');
      expect(avatars).toHaveLength(0);
    });

    it('匹配 single 模式快照', () => {
      const { container } = render(<ConverterSkeleton mode="single" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('模式: multi', () => {
    it('渲染多貨幣轉換器骨架', () => {
      render(<ConverterSkeleton mode="multi" />);

      // ✅ multi 模式有多個 role="status"，使用 getAllByRole
      const skeletons = screen.getAllByRole('status');
      expect(skeletons.length).toBeGreaterThan(0);
    });

    it('渲染多個 CurrencyCardSkeleton (4個)', () => {
      render(<ConverterSkeleton mode="multi" />);

      // ✅ multi 模式有 5 個 role="status"
      // 1 個來自 ConverterSkeleton，4 個來自 CurrencyCardSkeleton
      const statuses = screen.getAllByRole('status');
      expect(statuses).toHaveLength(5);
    });

    it('匹配 multi 模式快照', () => {
      const { container } = render(<ConverterSkeleton mode="multi" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('邊界情況', () => {
    it('處理 undefined mode (default single)', () => {
      // @ts-expect-error - 測試邊界情況
      const { container } = render(<ConverterSkeleton />);

      // ✅ 應該渲染 single 模式
      const avatars = container.querySelectorAll('.rounded-full');
      expect(avatars).toHaveLength(0);
    });
  });
});

describe('性能測試', () => {
  it('快速渲染不影響 TTI', () => {
    const start = performance.now();

    render(<SkeletonLoader />);

    const duration = performance.now() - start;

    // ✅ Lighthouse 優化目標: <50ms
    expect(duration).toBeLessThan(50);
  });

  it('所有組件渲染時間 <100ms', () => {
    const start = performance.now();

    render(
      <>
        <SkeletonLoader />
        <CurrencyCardSkeleton />
        <ConverterSkeleton mode="single" />
        <ConverterSkeleton mode="multi" />
      </>,
    );

    const duration = performance.now() - start;
    expect(duration).toBeLessThan(100);
  });
});
