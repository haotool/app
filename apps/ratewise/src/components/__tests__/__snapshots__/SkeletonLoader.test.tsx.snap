// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConverterSkeleton > 模式: multi > 匹配 multi 模式快照 1`] = `
<div
  class="space-y-4 animate-pulse"
  role="status"
>
  <div
    class="h-12 bg-slate-200 rounded-lg"
  />
  <div
    class="grid grid-cols-1 md:grid-cols-2 gap-4"
  >
    <div
      class="bg-white rounded-lg p-3 shadow-sm animate-pulse"
      role="status"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="flex items-center gap-2"
        >
          <div
            class="w-8 h-8 bg-slate-200 rounded-full"
          />
          <div
            class="space-y-1"
          >
            <div
              class="h-4 w-12 bg-slate-200 rounded"
            />
            <div
              class="h-3 w-20 bg-slate-200 rounded"
            />
          </div>
        </div>
        <div
          class="h-4 w-16 bg-slate-200 rounded"
        />
      </div>
    </div>
    <div
      class="bg-white rounded-lg p-3 shadow-sm animate-pulse"
      role="status"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="flex items-center gap-2"
        >
          <div
            class="w-8 h-8 bg-slate-200 rounded-full"
          />
          <div
            class="space-y-1"
          >
            <div
              class="h-4 w-12 bg-slate-200 rounded"
            />
            <div
              class="h-3 w-20 bg-slate-200 rounded"
            />
          </div>
        </div>
        <div
          class="h-4 w-16 bg-slate-200 rounded"
        />
      </div>
    </div>
    <div
      class="bg-white rounded-lg p-3 shadow-sm animate-pulse"
      role="status"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="flex items-center gap-2"
        >
          <div
            class="w-8 h-8 bg-slate-200 rounded-full"
          />
          <div
            class="space-y-1"
          >
            <div
              class="h-4 w-12 bg-slate-200 rounded"
            />
            <div
              class="h-3 w-20 bg-slate-200 rounded"
            />
          </div>
        </div>
        <div
          class="h-4 w-16 bg-slate-200 rounded"
        />
      </div>
    </div>
    <div
      class="bg-white rounded-lg p-3 shadow-sm animate-pulse"
      role="status"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="flex items-center gap-2"
        >
          <div
            class="w-8 h-8 bg-slate-200 rounded-full"
          />
          <div
            class="space-y-1"
          >
            <div
              class="h-4 w-12 bg-slate-200 rounded"
            />
            <div
              class="h-3 w-20 bg-slate-200 rounded"
            />
          </div>
        </div>
        <div
          class="h-4 w-16 bg-slate-200 rounded"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ConverterSkeleton > 模式: single > 匹配 single 模式快照 1`] = `
<div
  class="space-y-4 animate-pulse"
  role="status"
>
  <div
    class="grid grid-cols-1 md:grid-cols-2 gap-4"
  >
    <div
      class="h-14 bg-slate-200 rounded-lg"
    />
    <div
      class="h-14 bg-slate-200 rounded-lg"
    />
  </div>
  <div
    class="h-16 bg-purple-200 rounded-xl"
  />
</div>
`;

exports[`CurrencyCardSkeleton > 匹配設計快照 1`] = `
<div
  class="bg-white rounded-lg p-3 shadow-sm animate-pulse"
  role="status"
>
  <div
    class="flex items-center justify-between"
  >
    <div
      class="flex items-center gap-2"
    >
      <div
        class="w-8 h-8 bg-slate-200 rounded-full"
      />
      <div
        class="space-y-1"
      >
        <div
          class="h-4 w-12 bg-slate-200 rounded"
        />
        <div
          class="h-3 w-20 bg-slate-200 rounded"
        />
      </div>
    </div>
    <div
      class="h-4 w-16 bg-slate-200 rounded"
    />
  </div>
</div>
`;

exports[`SkeletonLoader > 基本渲染 > 渲染所有主要區塊 1`] = `
<div
  aria-live="polite"
  class="min-h-screen bg-slate-50 p-4 md:p-6"
  role="status"
>
  <div
    class="mx-auto max-w-4xl space-y-6"
  >
    <div
      class="animate-pulse"
    >
      <div
        class="h-8 w-48 bg-slate-200 rounded-lg mb-2"
      />
      <div
        class="h-4 w-32 bg-slate-200 rounded"
      />
    </div>
    <div
      class="flex gap-2 mb-4 animate-pulse"
    >
      <div
        class="h-10 w-32 bg-slate-200 rounded-lg"
      />
      <div
        class="h-10 w-32 bg-slate-200 rounded-lg"
      />
    </div>
    <div
      class="bg-white rounded-2xl shadow-lg p-6 space-y-4 animate-pulse"
    >
      <div
        class="grid grid-cols-1 md:grid-cols-2 gap-4"
      >
        <div
          class="space-y-2"
        >
          <div
            class="h-4 w-16 bg-slate-200 rounded"
          />
          <div
            class="h-12 bg-slate-200 rounded-lg"
          />
        </div>
        <div
          class="space-y-2"
        >
          <div
            class="h-4 w-16 bg-slate-200 rounded"
          />
          <div
            class="h-12 bg-slate-200 rounded-lg"
          />
        </div>
      </div>
      <div
        class="space-y-2"
      >
        <div
          class="h-4 w-20 bg-slate-200 rounded"
        />
        <div
          class="h-14 bg-slate-200 rounded-lg"
        />
      </div>
      <div
        class="bg-purple-50 rounded-xl p-4 space-y-2"
      >
        <div
          class="h-6 w-full bg-purple-200 rounded"
        />
        <div
          class="h-4 w-3/4 bg-purple-200 rounded"
        />
      </div>
      <div
        class="flex gap-2 flex-wrap"
      >
        <div
          class="h-10 w-20 bg-slate-200 rounded-lg"
        />
        <div
          class="h-10 w-20 bg-slate-200 rounded-lg"
        />
        <div
          class="h-10 w-20 bg-slate-200 rounded-lg"
        />
        <div
          class="h-10 w-20 bg-slate-200 rounded-lg"
        />
      </div>
    </div>
    <div
      class="bg-white rounded-2xl shadow-lg p-6 animate-pulse"
    >
      <div
        class="h-4 w-32 bg-slate-200 rounded mb-4"
      />
      <div
        class="h-64 bg-slate-200 rounded-lg"
      />
    </div>
    <div
      class="bg-white rounded-2xl shadow-lg p-6 animate-pulse"
    >
      <div
        class="h-4 w-24 bg-slate-200 rounded mb-4"
      />
      <div
        class="grid grid-cols-2 md:grid-cols-3 gap-2"
      >
        <div
          class="h-12 bg-slate-200 rounded-lg"
        />
        <div
          class="h-12 bg-slate-200 rounded-lg"
        />
        <div
          class="h-12 bg-slate-200 rounded-lg"
        />
      </div>
    </div>
  </div>
  <span
    class="sr-only"
  >
    載入匯率資料中...
  </span>
</div>
`;
